### ISSUE TITLE
cpu.architecture is not being set on vm creation when cloning

### ISSUE THREAD

**Issue by mattburchett** [bpg/terraform-provider-proxmox#1575]
**Describe the bug**
It looks like upon creation of a VM using the provider, that the CPU `architecture` is not getting set correctly when calling the API, even though I have it set in the configuration.

**To Reproduce**

1. Create a VM resource with cpu.architecture set.
2. Run apply again after the creation is done.

Please also provide a minimal Terraform configuration that reproduces the issue.

```hcl
resource "proxmox_virtual_environment_vm" "vms" {
  for_each = local.lrhq-pve-hosts

  name        = each.key
  description = "Managed by Terraform"

  node_name = each.value.target-node
  vm_id     = split(".", each.value.ip.address)[3]

  agent {
    enabled = true
  }

  clone {
    datastore_id = each.value.hardware.storage
    retries      = 10
    node_name    = "lrhq-pve"
    vm_id        = each.value.template
  }

  cpu {
    cores   = each.value.hardware.cores
    sockets = 1
    type    = "host"
		architecture = "x86_64"
  }

  disk {
    datastore_id = each.value.hardware.storage
    interface    = "virtio0"
    size         = each.value.hardware.disk_size
  }

    user_data_file_id = proxmox_virtual_environment_file.ubuntu_cloud_config[each.key].id
  }

  lifecycle {
    ignore_changes = [
      initialization[0].user_data_file_id
    ]
  }

  memory {
    dedicated = each.value.hardware.memory
  }

  network_device {}

  on_boot = true

  operating_system {
    type = "l26"
  }

  serial_device {}

  depends_on = [cloudflare_record.proxmox-lrhq-pve-dns, proxmox_virtual_environment_file.ubuntu_cloud_config]
}
```

**Expected behavior**
There should be no changes when you re-apply.

**Additional context**
It looks like this may be related to #1524. 

When setting the attribute via JSON API myself, I can see it when I do a `pvesh get`. In this example, both 105 and 121 VMIDs had the issue.

```bash
lrhq-pve ~ [0]#         pvesh get /nodes/lrhq-pve/qemu/105/config | grep arch
│ arch        │ x86_64                                                                                                   │
lrhq-pve ~ [0]#         pvesh get /nodes/lrhq-pve/qemu/121/config | grep arch
lrhq-pve ~ [1]#
```

When I enable debug/trace logging, I can see in the PUT request that `arch` is not passed. 

```
2024-10-05T04:53:49.680Z [DEBUG] provider.terraform-provider-proxmox_v0.66.1: Sending HTTP Request: Accept=application/json Content-Length=102 Content-Type=application/x-www-form-urlencoded Csrfpreventiontoken="6700C65D:WDF1cgPh+vym0GD87V/PH+NMB4cgT5m9OCoaVg+l2w4" tf_mux_provider=tf5to6server.v5tov6Server tf_http_trans_id=cbbf326d-5b01-bb68-5e67-8c7a45cd2d56 tf_resource_type=proxmox_virtual_environment_vm @module=proxmox Accept-Encoding=gzip tf_http_op_type=request tf_http_req_body="cores=4&cpu=cputype%3Dhost&cpuunits=1024&delete=vcpus%2Ccpulimit&name=lrhq-noble-test&numa=0&sockets=1" tf_http_req_uri=/api2/json/nodes/lrhq-pve/qemu/121/config tf_http_req_version=HTTP/1.1 tf_rpc=ApplyResourceChange @caller=/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.34.0/helper/logging/logging_http_transport.go:162 Cookie="_redacted_" Host=redacted:8006 tf_req_id=b63345fa-72f4-f18a-d40d-7344cd925da5 User-Agent=Go-http-client/1.1 tf_http_req_method=PUT tf_provider_addr=registry.terraform.io/bpg/proxmox timestamp=2024-10-05T04:53:49.679Z
```

In summary, the provider does not appear to correctly apply the cpu.architecture setting, causing it to be omitted in API requests despite being specified in the configuration. This results in Terraform showing the same changes on every plan, as the architecture setting was not sent to the API with the initial creation or the attempted modification.

- Single or clustered Proxmox: Single
- Proxmox version: 8.2.7
- Provider version (ideally it should be the latest version): latest (0.66.1)
- Terraform/OpenTofu version: Terraform 1.9.7
- OS (where you run Terraform/OpenTofu from): Ubuntu 22.04 LTS

**Comment by bpg** [comment]
I believe this only happens when cloning. Creating a VM from scratch does not show the same behavior

**Comment by mattburchett** [comment]
> I believe this only happens when cloning. Creating a VM from scratch does not show the same behavior

Hmm, I wonder if it isn't set on the template. I can confirm that if you'd like.

**Comment by bpg** [comment]
> > I believe this only happens when cloning. Creating a VM from scratch does not show the same behavior
> 
> Hmm, I wonder if it isn't set on the template. I can confirm that if you'd like.

Yeah, it would be interesting to see what is in the template. Handling defaults when cloning is tricky in the provider.

**Comment by mattburchett** [comment]
This is what I get out of pvesh about the template: 

```
#       pvesh get /nodes/lrhq-pve/qemu/9007/config
┌─────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ key         │ value                                                                                                   │
╞═════════════╪═════════════════════════════════════════════════════════════════════════════════════════════════════════╡
│ acpi        │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ agent       │ enabled=0,fstrim_cloned_disks=0,type=virtio                                                             │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ balloon     │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ bios        │ seabios                                                                                                 │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ boot        │ order=virtio0;net0                                                                                      │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cores       │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cpu         │ cputype=kvm64,flags=+aes                                                                                │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cpuunits    │ 1024                                                                                                    │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ description │ Managed by Terraform                                                                                    │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ digest      │ 4edd1d6847802a50efdfd9503e7e96eab8abba59                                                                │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ keyboard    │ en-us                                                                                                   │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ memory      │ 512                                                                                                     │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ name        │ ubuntu-noble-template                                                                                   │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ numa        │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ onboot      │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ ostype      │ other                                                                                                   │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ protection  │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ scsihw      │ virtio-scsi-pci                                                                                         │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ smbios1     │ uuid=93915516-a8fc-49c7-90d3-29e4a8cf905a                                                               │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ sockets     │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ tablet      │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ template    │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ vga         │ memory=16,type=std                                                                                      │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ virtio0     │ local-zfs:vm-9007-disk-0,aio=io_uring,backup=1,cache=none,discard=ignore,iothread=0,replicate=1,size=8G │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ vmgenid     │ a41cd6a4-2f4f-48c4-b769-5dff12a6d6ee                                                                    │
└─────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

I'm going to put the architecture flag on the template (which was also created around the time this VM was made) and see what that does. I am doing a full clone, not a linked clone or anything, so I figured I could set my own settings and not worry about the template too much.

I'll let you know what happens after I change that.

**Comment by mattburchett** [comment]
This is after the template was recreated, with that flag: 

```
pvesh get /nodes/lrhq-pve/qemu/9007/config
┌─────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ key         │ value                                                                                                   │
╞═════════════╪═════════════════════════════════════════════════════════════════════════════════════════════════════════╡
│ acpi        │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ agent       │ enabled=0,fstrim_cloned_disks=0,type=virtio                                                             │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ balloon     │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ bios        │ seabios                                                                                                 │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ boot        │ order=virtio0;net0                                                                                      │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cores       │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cpu         │ cputype=host                                                                                            │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cpuunits    │ 1024                                                                                                    │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ description │ Managed by Terraform                                                                                    │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ digest      │ 7bba3be49816323dbe2917d7f1511b8f206ff626                                                                │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ keyboard    │ en-us                                                                                                   │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ memory      │ 512                                                                                                     │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ name        │ ubuntu-noble-template                                                                                   │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ numa        │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ onboot      │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ ostype      │ other                                                                                                   │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ protection  │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ scsihw      │ virtio-scsi-pci                                                                                         │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ smbios1     │ uuid=bd069e48-34b5-476f-9eec-7d66697b6704                                                               │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ sockets     │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ tablet      │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ template    │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ vga         │ memory=16,type=std                                                                                      │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ virtio0     │ local-zfs:vm-9007-disk-0,aio=io_uring,backup=1,cache=none,discard=ignore,iothread=0,replicate=1,size=8G │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ vmgenid     │ 6f19ebe3-fc19-46f4-aa92-a7f136923607                                                                    │
└─────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

The resource, for context: 

```terraform
resource "proxmox_virtual_environment_vm" "ubuntu-noble-template" {
  name        = "ubuntu-noble-template"
  description = "Managed by Terraform"

  node_name = "lrhq-pve"
  vm_id     = 9007

  cpu {
    cores        = 1
    sockets      = 1
    type         = "host"
    architecture = "x86_64"
  }

  disk {
    datastore_id = "local-zfs"
    file_id      = proxmox_virtual_environment_download_file.ubuntu_noble_template.id
    interface    = "virtio0"
  }

  on_boot  = false
  started  = false
  template = true
}
```

And after tainting the test VM that I made to replicate the problem, letting terraform do the destroy and recreate off the new template, the flag was not carried across:

```bash
pvesh get /nodes/lrhq-pve/qemu/121/config
┌─────────────┬─────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ key         │ value                                                                                                   │
╞═════════════╪═════════════════════════════════════════════════════════════════════════════════════════════════════════╡
│ acpi        │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ agent       │ enabled=1,fstrim_cloned_disks=0,type=virtio                                                             │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ balloon     │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ bios        │ seabios                                                                                                 │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ boot        │ order=virtio0;net0                                                                                      │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cicustom    │ user=local:snippets/lrhq-noble-test-ubuntu.yml                                                          │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cores       │ 4                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cpu         │ cputype=host                                                                                            │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ cpuunits    │ 1024                                                                                                    │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ description │ Managed by Terraform                                                                                    │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ digest      │ 372cedaff242918e350825e7af0c82fb5abe5df0                                                                │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ ide2        │ local-zfs:vm-121-cloudinit,media=cdrom                                                                  │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ ipconfig0   │ gw=10.90.10.1,ip=10.90.15.121/17                                                                        │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ keyboard    │ en-us                                                                                                   │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ memory      │ 2048                                                                                                    │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ name        │ lrhq-noble-test                                                                                         │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ net0        │ virtio=BC:24:11:98:DE:F5,bridge=vmbr0,firewall=0                                                        │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ numa        │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ onboot      │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ ostype      │ l26                                                                                                     │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ protection  │ 0                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ scsihw      │ virtio-scsi-pci                                                                                         │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ serial0     │ socket                                                                                                  │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ smbios1     │ uuid=ee8c8f4e-3fc1-46a0-9716-97b4fbe4b8b3                                                               │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ sockets     │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ tablet      │ 1                                                                                                       │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ vga         │ memory=16,type=std                                                                                      │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ virtio0     │ local-zfs:vm-121-disk-0,aio=io_uring,backup=1,cache=none,discard=ignore,iothread=0,replicate=1,size=20G │
├─────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ vmgenid     │ ef8ce663-e901-41ac-927f-9c76cb3345c8                                                                    │
└─────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────┘
```

**Comment by 1-cameron** [comment]
I'm experiencing similar behaviour but with a VM that's been restored (originally templated) from backup before being imported - `terraform apply` will reboot the VM and `architecture` doesn't get set in the state.

**Comment by ratiborusx** [comment]
> I'm experiencing similar behaviour both with clones & a VM that's been restored (originally cloned from a template) from backup before being imported to the state - `terraform apply` will run but `architecture` doesn't get set in the state.

You're probably working under non-root account. You may check my mad ramblings about that in #1494 but proceed with caution.

**Comment by mattburchett** [comment]
> > I'm experiencing similar behaviour both with clones & a VM that's been restored (originally cloned from a template) from backup before being imported to the state - `terraform apply` will run but `architecture` doesn't get set in the state.
> 
> You're probably working under non-root account. You may check my mad ramblings about that in #1494 but proceed with caution.

Correct, I am using a terraform service account rather than root, due to some commands requiring `bash` instead of my default shell as the root user. 

Thanks for the information in the other issue, I'm glad you found the source of the problem. 🙂
